contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -888,18 +923,287 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
 	return new_candidates;
 }
 
+/*
+ * get_immediate_base_type_of_UDT_internal()
+ * This function returns the Immediate base type for UDT.
+ * Returns InvalidOid if given type is not an UDT
+ */
+static Oid
+get_immediate_base_type_of_UDT_internal(Oid typeid)
+{
+	HeapTuple					tuple;
+	bool						isnull;
+	Datum						datum;
+	Datum                       tsql_typename;
+	Oid							base_type;
+	LOCAL_FCINFO(fcinfo, 1);
+
+	/* if common_utility_plugin_ptr is not initialised */
+	if (common_utility_plugin_ptr == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+					errmsg("Failed to find common utility plugin.")));
+
+	/* if tsql_typename is NULL it implies that inputTypId corresponds to UDT */
+	InitFunctionCallInfoData(*fcinfo, NULL, 0, InvalidOid, NULL, NULL);
+	fcinfo->args[0].value = ObjectIdGetDatum(typeid);
+	fcinfo->args[0].isnull = false;
+	tsql_typename = (*common_utility_plugin_ptr->translate_pg_type_to_tsql) (fcinfo);
+
+	/* if given type is not an UDT then return InvalidOid */
+	if (tsql_typename)
+		return InvalidOid;
+
+	/* Get immediate base type id of given type id */
+	tuple = SearchSysCache1(TYPEOID, ObjectIdGetDatum(typeid));
+	if (!HeapTupleIsValid(tuple))
+		return InvalidOid;
+
+	datum = SysCacheGetAttr(TYPEOID, tuple, Anum_pg_type_typbasetype, &isnull);
+	if (isnull)
+		return InvalidOid;
+
+	base_type = DatumGetObjectId(datum);
+	ReleaseSysCache(tuple);
+
+	return base_type;
+}
+
+Datum
+get_immediate_base_type_of_UDT(PG_FUNCTION_ARGS)
+{
+	Oid			base_type;
+	
+	base_type = get_immediate_base_type_of_UDT_internal(PG_GETARG_OID(0));
+	if (!OidIsValid(base_type))
+		PG_RETURN_NULL();
+
+	PG_RETURN_OID(base_type);
+}
+
+void
+init_special_function_list()
```

Comment:
We are trying to call this during initialisation of tsql extension which may increase overall time to make a new connection to database. Instead, did we consider the approach to load this on demand?

Reply:
Yes previously we were using the approach of loading the typeid on demand, but @rishabhtanwar29 suggested [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2495#discussion_r1672638837) to load it in pg_init instead.

cc: @rishabhtanwar29

Reply:
Hash table initialization should better happen once during library load to save the extra efforts later. Regarding the increase in init time, this hash table is anyway getting initialized just after initialization of tsql_precedence hash table which would preload all of the datatype OIDs into syscache so this step is anyway not going to take much time.
Additionally, we do have logic still to load this table on demand in case the load was incomplete previously.